### PR TITLE
Add trace_dot in linalg

### PR DESCRIPTION
**Pull request by vinx13**
* Add method that computes trace of the product of two matrixes in linalg
* Use `trace_dot` in LMNN

### CODE REVIEW

File: src/shogun/metric/LMNN.cpp [hunk 1]
```diff
@@ -104,13 +104,10 @@ void CLMNN::train(SGMatrix<float64_t> init_transform)
...
 		// Compute the objective, trace of Mahalanobis distance matrix (L squared) times the gradient
 		// plus the number of current impostors to account for the margin
-		// trace of the product of two matrices computed fast using
-		// trace(A*B)=sum(A.*B')
 		SG_DEBUG("Computing objective.\n")
 		obj[iter] = m_regularization * cur_impostors.size();
-		obj[iter] += linalg::trace(
-		    linalg::element_prod(
-		        linalg::matrix_prod(L, L, true, false), gradient));
+		obj[iter] +=
```
Comment (karlnapf) [inline src/shogun/metric/LMNN.cpp]: Is this the same as before? Doesnt seem to me but maybe I am blind :)

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 2]
```diff
@@ -1681,6 +1681,19 @@ namespace shogun
...
+		 * Method that computes the trace of \f$AB\f$ as \f$sum(A.*B')$\f
+		 *
+		 * @param A The matrix A
+		 * @param B The matrix B
+		 * @return The trace of the product of A and B
+		 */
+		template <typename T>
+		T trace_dot(const SGMatrix<T>& A, const SGMatrix<T>& B)
+		{
+			return sum(element_prod(A, transpose_matrix(B)));
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: is transpose matrix changing the memory (it shouldnt)

Also, the API seems inconsistent, for some methods we pass flags for whether matrices should be transposed. And for others, this flag doesnt exist and we have to transpose matrices from the outside ?

File: src/shogun/metric/LMNN.cpp (hunk 1 above)
Comment (vinx13) [reply src/shogun/metric/LMNN.cpp]: They are different. Actually I was wrong in #4167. Now it is the same as the original one.  https://github.com/shogun-toolbox/shogun/blob/431e56b67344e0662444c0efbd4aefde47990fa4/src/shogun/metric/LMNN.cpp#L113

File: src/shogun/mathematics/linalg/LinalgNamespace.h (hunk 2 above)
Comment (vinx13) [reply src/shogun/mathematics/linalg/LinalgNamespace.h]: transpose matrix creates a new matrix.  
The API is indeed inconsistent, maybe we should add transpose flags to element_prod ?

Comment (karlnapf) [reply src/shogun/mathematics/linalg/LinalgNamespace.h]: yep!

File: src/shogun/metric/LMNN.cpp (hunk 1 above)
Comment (karlnapf) [reply src/shogun/metric/LMNN.cpp]: EHM!!!
Why did the tests pass? Or are there none?

Comment (vinx13) [reply src/shogun/metric/LMNN.cpp]: Well, here the trace is used as the termination condition. Seems that it doesn't stop until reaching maximum number of iterations when trace is wrong in tests.

Comment (karlnapf) [reply src/shogun/metric/LMNN.cpp]: I see now! makes sense
Ok, this definitely should be tested somehow (at least in a unit test)
Wanna do send another PR for that?

**Comment by karlnapf** [comment]
What is the status of this one then? :)

File: src/shogun/mathematics/linalg/LinalgNamespace.h (hunk 2 above)
Comment (vinx13) [reply src/shogun/mathematics/linalg/LinalgNamespace.h]: @karlnapf I made another pr #4183 , with which I could make it clear here

**Comment by vinx13** [comment]
This one would be ready for review after cleaning up calls to transpose_matrix.

**Comment by karlnapf** [comment]
ok ping me here once that is the case

**Comment by vinx13** [comment]
@karlnapf Ready for review now :)

File: src/shogun/metric/LMNN.cpp (hunk 1 above)
Comment (karlnapf) [reply src/shogun/metric/LMNN.cpp]: integration tests are back, so now let's see :)

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 3]
```diff
@@ -1734,6 +1734,23 @@ TEST(LinalgBackendEigen, SGMatrix_trace)
 	EXPECT_NEAR(trace(A), tr, 1e-15);
 }
 
+TEST(LinalgBackendEigen, SGMatrix_tract_dot)
```
Comment (karlnapf) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: typo

**Comment by karlnapf** [comment]
typo fix and travis green and we can merge this :)

**Comment by karlnapf** [comment]
restarted the hickup builds...should be fine soon

File: src/shogun/metric/LMNN.cpp (hunk 1 above)
Comment (iglesias) [reply src/shogun/metric/LMNN.cpp]: Hey guys. Why isn't the same as before? I don't see it :-/ Or maybe this discussion is about a previous diff?

Comment (karlnapf) [reply src/shogun/metric/LMNN.cpp]: Check the history of that line, it changed back and forth IIRC

Comment (iglesias) [reply src/shogun/metric/LMNN.cpp]: I checked and only found the three mentioned here: the original with TRACE plust the two in this diff.

Comment (vinx13) [reply src/shogun/metric/LMNN.cpp]: The original with TRACE is correct, the former one in this diff is wrong.

Comment (iglesias) [reply src/shogun/metric/LMNN.cpp]: Where is the difference between them, @vinx13?

Comment (vinx13) [reply src/shogun/metric/LMNN.cpp]: We are calculating trace of L' * L * gradient, which is sum( (L'*L) .* gradient' ), instread of trace( (L'*L) .* gradient) that appears in the previous commit

**Pull request merged**